clients/go/zbc/client_test.go

```diff
@@ -0,0 +1,139 @@
+// Copyright © 2018 Camunda Services GmbH (info@camunda.com)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package zbc
+
+import (
+	"fmt"
+	"github.com/stretchr/testify/require"
+	"github.com/zeebe-io/zeebe/clients/go/pb"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/status"
+	"net"
+	"strings"
+	"testing"
+)
+
+func TestNewZBClientWithTls(t *testing.T) {
+	lis, grpcServer := createSecureServer()
+
+	go grpcServer.Serve(lis)
+	defer func() {
+		grpcServer.Stop()
+		_ = lis.Close()
+	}()
+
+	parts := strings.Split(lis.Addr().String(), ":")
+	client, e := NewZBClient(&ZBClientConfig{
+		GatewayAddress:      fmt.Sprintf("0.0.0.0:%s", parts[len(parts)-1]),
+		UseSecureConnection: true,
+		CaCertificatePath:   "../resources/ca.cert.pem",
+	})
+
+	require.NoError(t, e)
+
+	_, err := client.NewTopologyCommand().Send()
+
+	require.Error(t, err)
+	if status, ok := status.FromError(err); ok {
+		require.Equal(t, codes.Unimplemented, status.Code())
+	}
+}
+
+func TestNewZBClientWithoutTls(t *testing.T) {
+	lis, _ := net.Listen("tcp", "0.0.0.0:0")
+
+	grpcServer := grpc.NewServer()
+	pb.RegisterGatewayServer(grpcServer, &pb.UnimplementedGatewayServer{})
+
+	go grpcServer.Serve(lis)
+	defer func() {
+		grpcServer.Stop()
+		_ = lis.Close()
+	}()
+
+	parts := strings.Split(lis.Addr().String(), ":")
+	client, e := NewZBClient(&ZBClientConfig{
+		GatewayAddress:      fmt.Sprintf("0.0.0.0:%s", parts[len(parts)-1]),
+		UseSecureConnection: false,
+		CaCertificatePath:   "../resources/ca.cert.pem",
+	})
+
+	require.NoError(t, e)
+
+	_, err := client.NewTopologyCommand().Send()
+
+	require.Error(t, err)
+	if status, ok := status.FromError(err); ok {
+		require.Equal(t, codes.Unimplemented, status.Code())
+	}
+
+}
+
+func TestNewZBClientWithDefaultRootCa(t *testing.T) {
+	lis, grpcServer := createSecureServer()
+
+	go grpcServer.Serve(lis)
```

Comment:
Isn't the setup always kind of the same? Can we make this reusable?

Reply:
Do you mean the `go grpcServer.Serve(lis)`? I thought it would be more obvious what the deferred is undoing if the `Server` call was next to it but I can move it inside `createSecureServer` if you prefer

Reply:
Not the defer, I meant more generally most of the tests are the same minus a few params, but looking at it again I don't think it's worth the effort to make it more generic. It's good as is :+1: